[HLSL] cbuffer: Create host layout structs

hekota opened the pull request:
Creates layout struct for `cbuffer` in Sema which will contains only declarations contributing to the constant buffer layout. Anything else will be filtered out, such as static variables decls, struct and function definitions, resources, or empty struct and zero-sized arrays.

If the constant buffer includes a struct that contains any of the above undesirable declarations, a new version of this struct should be created with these declarations filtered out as well.

The definition of buffer layout struct will be added to the HLSLBufferDecl AST node as the last node. Any layout structs for embedded structures will be added there as well.

Fixes #122553

hekota committed:
[HLSL] cbuffer: Create host layout struct and add resource handle to AST

Creates layout struct for `cbuffer` in Sema which will contains only declarations
contributing to the constant buffer layout. Anything else will be filtered out,
such as static variables decls, struct and function definitions, resources,
or empty struct and zero-sized arrays.

If the constant buffer includes a struct that contains any of the above undesirable
declarations, a new version of this struct should be created with these declarations
filtered out as well.

The definition of buffer layour struct is added to the HLSLBufferDecl node and is followed
by 'cbuffer` resource handle decl referencing the layout struct as its contained type.

Fixes #122553

hekota committed:
Merge branch 'main' into cbuffer-layout-structs

llvm-beanz commented on clang/lib/Sema/SemaHLSL.cpp:
```diff
@@ -253,12 +257,253 @@ static void validatePackoffset(Sema &S, HLSLBufferDecl *BufDecl) {
+      createHostLayoutStructForBuffer(SemaRef, BufDecl);
+
+  // create buffer resource handle
+  createHLSLBufferHandle(SemaRef, BufDecl, LayoutStruct);
```
Do we need to put the handle into the layout structure? That seems odd and unintuitive to me, and it seems like we filter it out again in a few places.

hekota replied:
The handle is not added into the layout structure, it is added into the 
HLSLBufferDecl.

hekota replied:
We don't need to create the handle decl in the front end, we can certainly do it later in codegen.

llvm-beanz reviewed (commented):
The one design question I have here is: should we be bothering to create a handle for the cbuffer in the frontend at all?

My thought had been that we would instead emit the cbuffer member declarations as constant address space (2 for DXIL) declarations and have a pass to collect them and replace them with accesses off a handle that we generate in the backend.

Then we just need to generate metadata in the frontend for the buffer's contained type, and which global variables are bound to which cbuffer.

hekota committed:
Remove resource handle from HLSLBufDecl, cleanup tests

hekota committed:
One more char to revert things to previous state

hekota commented:
> My thought had been that we would instead emit the cbuffer member declarations as constant address space (2 for DXIL) declarations and have a pass to collect them and replace them with accesses off a handle that we generate in the backend.
> 
> Then we just need to generate metadata in the frontend for the buffer's contained type, and which global variables are bound to which cbuffer.

I was kind of hoping we might be able to generate the accesses off a handle during codegen when the global is accessed. I'm not sure if that's possible though.

llvm-beanz commented:
> I was kind of hoping we might be able to generate the accesses off a handle during codegen when the global is accessed. I'm not sure if that's possible though.

I'm concerned that this will be more complex to make work with the other CBV syntaxes. For example, the implicit globals don't have a bufferdecl to attach to, so we'll need an extra implicit decl to attach them to, and the `ConstantBuffer` type has a similar problem.

If instead we treat CBV as a storage class with an associated address space (similar to `groupshared`), we can emit cbuffer elements as globals in that address space, standalone globals do the same, and `ConstantBuffer<T>` can effectively be an alias something like: `template <typename T> using ConstantBuffer = constant T`, where the `constant` storage class specifier denotes that it is a unique CBV declaration. This would also solve the problem of `ConstantBuffer`'s member expression operator lookup.

hekota committed:
Make the layout struct name shorter

Use "__layout." prefix instead of "__hostlayout.struct." to make the text output more concise for both AST and LLVM IR.

llvm-beanz commented on clang/lib/Sema/SemaHLSL.cpp:
```diff
@@ -253,12 +257,227 @@ static void validatePackoffset(Sema &S, HLSLBufferDecl *BufDecl) {
+  unsigned suffix = 0;
+  while (true) {
+    if (suffix != 0)
+      II = &AST.Idents.get((llvm::Twine(Name) + "." + Twine(suffix)).str(),
```
I wonder if it would be easier to test if these types were spellable? I could imagine something like:

```
cbuffer CB {
  int X;
  RWBuffer B;
}

struct CBStruct {
  int X;
};
_Static_assert(__builtin_hlsl_is_scalarized_layout_compatible(CBStruct, __layout_CB), "");
```

Feel free to ignore if you don't think this is useful, but it might be easier to deal with than the AST FileCheck tests.

hekota replied:
This is an interesting way to test this! I like seeing what the generated AST looks like, but I will add the static asserts as well as an additional verification. I will need to change the `__layout*` struct name to be a valid identifier though, currently it is using `.` in the name ;).

llvm-beanz reviewed (approved):
This all looks pretty good to me. I have a few small comments and suggestions, but nothing that needs to block this from going in.

hekota committed:
code review feedback - rename function and layout struct, add static asserts

tex3d commented on clang/lib/Sema/SemaHLSL.cpp:
```diff
@@ -253,12 +257,229 @@ static void validatePackoffset(Sema &S, HLSLBufferDecl *BufDecl) {
+    MustBeUnique = true;
+  }
+
+  std::string Name = "__layout_" + NameBase;
```
Could we use a prefix/suffix that is unspellable in HLSL syntax?  Such as a `.host_layout` suffix or something?

farzonl commented on clang/lib/Sema/SemaHLSL.cpp:
```diff
@@ -253,12 +257,229 @@ static void validatePackoffset(Sema &S, HLSLBufferDecl *BufDecl) {
+                                               bool MustBeUnique,
+                                               DeclContext *DC) {
+  ASTContext &AST = S.getASTContext();
+  std::string NameBase;
```
many of the other functions in this file are using l`lvm::StringRef`. Why are we not doing the same thing for this function?

hekota replied:
I can change `NameBase` to be `StringRef`, but in other places where this function it is creating new strings by appending the `StringRef` does not work, right?

farzonl replied:
Just wanted to make sure `StringRef ` was considered and this was intentional. no change needed.

llvm-beanz replied:
`llvm::SmallString` might be a better option to reduce the likelihood of needing an allocation, and unifying `NameBase` and `Name` to a single `SmallString` would potentially reduce dynamic allocations and copying of string data.

tex3d commented on clang/lib/Sema/SemaHLSL.cpp:
```diff
@@ -253,12 +257,229 @@ static void validatePackoffset(Sema &S, HLSLBufferDecl *BufDecl) {
+  LookupResult R(S, NameInfo, Sema::LookupOrdinaryName);
+  S.LookupName(R, S.getScopeForContext(DC));
+  if (R.isSingleResult())
+    return R.getAsSingle<CXXRecordDecl>();
```
Could this be multiple results if the name is defined in multiple accessible scopes (like current and parent scopes)?  Or does it only return zero or one result from the specified scope?

hekota replied:
This should only return zero or one result in the specified scope, which for the `HLSLBufferDecl` is the global scope. I will update the `LookupNameKind` to `LookupTagName` though to be more specific that the lookup is for a record/struct.

tex3d replied:
Ok, I don't see cases like this tested:
```hlsl
namespace NS1 {
  struct Foo { ... };
  struct Bar { struct Foo { ... }; ... };
}
struct Foo { ... };
...
cbuffer CB1 {
  Foo foo1;
  NS1::Foo foo2;
  NS1::Bar::Foo foo3;
}

namespace NS2 {
  struct Foo { ... };
  cbuffer CB2 {
    ::Foo foo0;
    Foo foo1;
    NS1::Foo foo2;
    NS1::Bar::Foo foo3;
  }
}
```

In what decl context will implicit host layout decls be created for each type?  Will there be duplicates?

hekota replied:
I see what you mean now. I have added a separate cbuffer AST test with namespaces and changed the implementation to always create the layout struct on the same context as the original struct so that it will be in the same namespace. The lookup is needs to scan just the immediate non-transparent decl context.

farzonl commented on clang/lib/Sema/SemaHLSL.cpp:
```diff
@@ -253,12 +257,229 @@ static void validatePackoffset(Sema &S, HLSLBufferDecl *BufDecl) {
+  return HLSLAttributedResourceType::findHandleTypeOnResource(Ty) != nullptr;
+}
+
+static CXXRecordDecl *createHostLayoutStruct(Sema &S, CXXRecordDecl *StructDecl,
```
NIT: I usually like to put function decls at the top of the file before any function definitions to give me more freedom on code placement/movement.

farzonl commented on clang/lib/Sema/SemaHLSL.cpp:
```diff
@@ -253,12 +257,229 @@ static void validatePackoffset(Sema &S, HLSLBufferDecl *BufDecl) {
+        return nullptr;
+      Ty = RD->getTypeForDecl();
+    }
+  } else if (Ty->isConstantArrayType()) {
```
```suggestion
  }
  if (Ty->isConstantArrayType() && isZeroSizedArray(cast<ConstantArrayType>(Ty))) {
    return nullptr;
}
```

tex3d commented on clang/lib/Sema/SemaHLSL.cpp:
```diff
@@ -253,12 +257,229 @@ static void validatePackoffset(Sema &S, HLSLBufferDecl *BufDecl) {
+                                                 CXXRecordDecl *LayoutStruct,
+                                                 HLSLBufferDecl *BufDecl) {
+  if (Ty->isRecordType()) {
+    if (isResourceRecordType(Ty))
```
Won't we want to catch more intangible types than just resources here?  How do we make sure this is kept up-to-date for all intangible types?

On a related note, how do we ensure this code under `createHostLayoutStruct` always has matching/consistent logic to the code in `requiresImplicitBufferLayoutStructure`?  Could there be a way to share the key logic components perhaps?  For instance, there could be functions that return whether a leaf type is intangible (maybe this is just `Type::IsHLSLIntangibleType`), or is zero-sized, separate from the logic that recursively traverses a record type looking for these properties on field types.  The `IsHLSLIntangibleType` and zero-sized detection functions are then used by both traversal paths to ensure consistent behavior.

tex3d replied:
Ok, I don't see cases like this tested:
```hlsl
namespace NS1 {
  struct Foo { ... };
  struct Bar { struct Foo { ... }; ... };
}
struct Foo { ... };
...
cbuffer CB1 {
  Foo foo1;
  NS1::Foo foo2;
  NS1::Bar::Foo foo3;
}

namespace NS2 {
  struct Foo { ... };
  cbuffer CB2 {
    ::Foo foo0;
    Foo foo1;
    NS1::Foo foo2;
    NS1::Bar::Foo foo3;
  }
}
```

In what decl context will implicit host layout decls be created for each type?  Will there be duplicates?

hekota replied:
There is an assert at the top of the `createHostLayoutStruct` function that will make sure the logic is consistent: 
```
assert(requiresImplicitBufferLayoutStructure(StructDecl) && "struct is already HLSL buffer compatible");
```

I have also moved the checking into a shared function `isInvalidConstantBufferLeafElementType`. Currently the only intangible types are resources classes, but I added a check for builtin resource type as well.

hekota committed:
code review feedback - use StringRef, update lookup kind

llvm-beanz commented on clang/lib/Sema/SemaHLSL.cpp:
```diff
@@ -253,12 +257,229 @@ static void validatePackoffset(Sema &S, HLSLBufferDecl *BufDecl) {
+
+  // copy base struct, create HLSL Buffer compatible version if needed
+  if (unsigned NumBases = StructDecl->getNumBases()) {
+    assert(NumBases == 1 && "HLSL supports only one base type");
```
I don't think it is accurate to say HLSL supports multiple bases. The `interface` support in DXC is HLSL 2015-only, we don't do anything sane with it in the language versions that DXC actually supports compiling for.

tex3d replied:
> The interface support in DXC is HLSL 2015-only

Not entirely true.  We don't support use of interface instances (basically like pointers), but we do support interface definitions, inheritance from such, and overriding method implementations.

There is a bug in DXC when calling a method originally defined in an interface that is not inherited first in a multiple-inheritance scenario due to the way the MicrosoftCXXABI generates the code to access `this` for a method which we inherited by defaulting to that ABI in DXC.

hekota replied:
Ok, so there can be multiple interfaces, but only one base struct, correct? And interfaces have only methods and no data members, so they can be safely ignored for the layout struct.

hekota replied:
And Clang does not support interfaces yet.

hekota replied:
I have filed an [issue](https://github.com/llvm/llvm-project/issues/124178) and added a FIXME to the code.

hekota replied:
It looks like we might not be supporting interfaces in Clang, so I'll remove the FIXME.

hekota committed:
Code review feedback
- create common function isInvalidConstantBufferLeafElementType
- create layout structs in the same context as the original struct
- fix lookup to scan just the first non-transparent context
- add test case with namespaces

hekota committed:
clang-format

hekota committed:
Use llvm::SmallString, add FIXME about interfaces

hekota committed:
remove FIXME
